File: src/main/java/com/janrain/backplane/server1/BusConfig1.java
/*
 * Copyright 2012 Janrain, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.janrain.backplane.server1;

import com.janrain.backplane.common.BackplaneServerException;
import com.janrain.backplane.common.ExternalizableCore;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;
import org.apache.commons.lang.StringUtils;
import org.apache.log4j.Logger;

import java.util.*;

/**
 * @author Johnny Bufu
 */
public class BusConfig1 extends ExternalizableCore {

    // - PUBLIC

    public BusConfig1() {}

    public BusConfig1(String busName, String owner, String retentionTimeSeconds, String retentionTimeStickySeconds) throws BackplaneServerException {
        Map<String,String> d = new LinkedHashMap<String, String>();
        d.put(Field.BUS_NAME.getFieldName(), busName);
        d.put(Field.RETENTION_TIME_SECONDS.getFieldName(), retentionTimeSeconds);
        d.put(Field.RETENTION_STICKY_TIME_SECONDS.getFieldName(), retentionTimeStickySeconds);
        d.put(owner, BUS_PERMISSION.GETALL.name());
        try {
            init(busName, d);
        } catch (MessageException e) {
            throw new BackplaneServerException(e.getMessage(), e);
        }
    }

    public enum BUS_PERMISSION { GETALL, POST, GETPAYLOAD, IDENTITY }

    @Override
    public String getIdValue() {
        return get(Field.BUS_NAME);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(Field.class);
    }

    public String getBusName() {
        return getIdValue();
    }

    public int getRetentionTimeSeconds() {
        return Integer.valueOf(get(Field.RETENTION_TIME_SECONDS));
    }

    public int getRetentionTimeStickySeconds() {
        return Integer.valueOf(get(Field.RETENTION_STICKY_TIME_SECONDS));
    }

    public Collection<BUS_PERMISSION> getPermissions(String user) {
        if (isBusConfigField(user)) {
            throw new IllegalArgumentException("Invalid user name: " + user);
        }

        String perms = get(user);
        EnumSet<BUS_PERMISSION> result = EnumSet.noneOf(BUS_PERMISSION.class);
        if (StringUtils.isNotBlank(perms)) {
            for(String perm : perms.split(",")) {
                result.add(BUS_PERMISSION.valueOf(perm));
            }
        }
        return result;
    }

    public static enum Field implements MessageField {

        BUS_NAME,

        RETENTION_TIME_SECONDS {
            @Override
            public void validate(String value) throws MessageException {
                if (isRequired() || value != null) {
                    String fieldName = getFieldName();
                    int intValue = validateInt(fieldName, value);
                    if (intValue < RETENTION_MIN_SECONDS || intValue > RETENTION_MAX_SECONDS) {
                        throw new IllegalArgumentException("Value of " + fieldName + " = " + intValue + " but must be between " + RETENTION_MIN_SECONDS + " and " + RETENTION_MAX_SECONDS);
                    }
                }
            }},

        RETENTION_STICKY_TIME_SECONDS {
            @Override
            public void validate(String value) throws MessageException {
                if (isRequired() || value != null) {
                    String fieldName = getFieldName();
                    validateInt(fieldName, value);
                    int intValue = validateInt(fieldName, value);
                    if (intValue < RETENTION_STICKY_MIN_SECONDS || intValue > RETENTION_STICKY_MAX_SECONDS) {
                        throw new IllegalArgumentException("Value of " + fieldName + " = " + intValue + " but must be between " + RETENTION_STICKY_MIN_SECONDS + " and " + RETENTION_STICKY_MAX_SECONDS);
                    }
                }
            }};




        @Override
        public String getFieldName() {
            return name();
        }

        @Override
        public boolean isRequired() {
            return true;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(name(), value);
        }

        // - PRIVATE

        private static final int RETENTION_MIN_SECONDS = 60;
        private static final int RETENTION_MAX_SECONDS = 604800; // one week
        private static final int RETENTION_STICKY_MIN_SECONDS = 28800; // eight hours
        private static final int RETENTION_STICKY_MAX_SECONDS = 604800; // one week

    }


    // - PRIVATE

    private static final long serialVersionUID = 2634562172424519254L;

    private static final Logger logger = Logger.getLogger(BusConfig1.class);

    private boolean isBusConfigField(String name) {
        try {
            Field.valueOf(name);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
}

File: src/main/scala/com/janrain/backplane/server1/model/BusConfig1.scala
package com.janrain.backplane.server1.model

import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}
import com.janrain.backplane.common.MessageException
import scala.collection.JavaConversions._
import com.janrain.backplane.server.BusConfig1.BUS_PERMISSION
import com.janrain.backplane.dao.LegacySupport

/**
 * @author Johnny Bufu
 */
class BusConfig1(data: Map[String,String]) extends Message(data, BusConfig1Fields.values) with LegacySupport[com.janrain.backplane.server.BusConfig1] {

  def this(javaData: java.util.Map[String,String]) = this(BusConfig1.fromLegacy(javaData))

  def idField = BusConfig1Fields.BUS_NAME

  def isAllowed(user: String, permissionField: BusConfig1Fields.EnumVal) = get(permissionField).exists(_.split(",").contains(user))

  def retentionTimeSeconds: Int = get(BusConfig1Fields.RETENTION_TIME_SECONDS).getOrElse(BusConfig1.RETENTION_MIN_SECONDS.toString).toInt

  def retentionTimeStickySeconds: Int = get(BusConfig1Fields.RETENTION_STICKY_TIME_SECONDS).getOrElse(BusConfig1.RETENTION_STICKY_MIN_SECONDS.toString).toInt

  def asLegacy = {
    val getallUsers = get(BusConfig1Fields.GETALL_USERS).map(_.split(",")).toIterable.flatten
    val postUsers = get(BusConfig1Fields.POST_USERS).map(_.split(",")).toIterable.flatten

    val g = getallUsers.map(user => if (postUsers.contains(user)) user -> "GETALL,POST" else user -> "GETALL").toMap
    val p = postUsers.map(user => if (getallUsers.contains(user)) user -> "GETALL,POST" else user -> "POST").toMap
    val userPermissionsMap = g ++ p

    val legacyFieldNames = com.janrain.backplane.server.BusConfig1.Field.values().map(_.getFieldName.toLowerCase)

    val filtered = this.filterKeys(legacyFieldNames.contains(_)).map {
      case (k,v) => k.toUpperCase -> v
    }

    new com.janrain.backplane.server.BusConfig1(mapAsJavaMap(filtered ++ userPermissionsMap))
  }
}

object BusConfig1 {
  private[model] final val RETENTION_MIN_SECONDS = 60L
  private[model] final val RETENTION_MAX_SECONDS = 604800L        // one week
  private[model] final val RETENTION_STICKY_MIN_SECONDS = 28800L  // eight hours
  private[model] final val RETENTION_STICKY_MAX_SECONDS = 604800L // one week

  def fromLegacy(javaData: java.util.Map[String, String]): Map[String,String] = {
    val scalaMap = javaData.toMap
    val getallUsers = scalaMap.collect {
      case (k, v) if v.contains(BUS_PERMISSION.GETALL.name()) => k
    }
    val postUsers = scalaMap.collect {
      case (k, v) if v.contains(BUS_PERMISSION.POST.name()) => k
    }

    scalaMap.collect {
      case (k, v) if !getallUsers.contains(k) && !postUsers.contains(k) => k.toLowerCase -> v
    } ++
      Map( BusConfig1Fields.POST_USERS.name -> postUsers.mkString(","),
           BusConfig1Fields.GETALL_USERS.name -> getallUsers.mkString(","))
  }
}

object BusConfig1Fields extends MessageFieldEnum {

  type BusConfig1Field = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
  }

  val BUS_NAME = new BusConfig1Field { def name = "bus_name" }

  val RETENTION_TIME_SECONDS = new BusConfig1Field { def name = "retention_time_seconds"

    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateLong(fieldValue)
      fieldValue.map(_.toLong).foreach(longValue =>
        if (longValue < BusConfig1.RETENTION_MIN_SECONDS || longValue > BusConfig1.RETENTION_MAX_SECONDS)
          throw new MessageException("Value of " + name + " = " + longValue + " but must be between " + BusConfig1.RETENTION_MIN_SECONDS + " and " + BusConfig1.RETENTION_MAX_SECONDS)
      )
    }
  }

  val RETENTION_STICKY_TIME_SECONDS = new BusConfig1Field { def name = "retention_sticky_time_seconds"

    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateLong(fieldValue)
      fieldValue.map(_.toLong).foreach(longValue =>
        if (longValue < BusConfig1.RETENTION_STICKY_MIN_SECONDS || longValue > BusConfig1.RETENTION_STICKY_MAX_SECONDS)
          throw new MessageException("Value of " + name + " = " + longValue + " but must be between " + BusConfig1.RETENTION_MIN_SECONDS + " and " + BusConfig1.RETENTION_STICKY_MAX_SECONDS)
      )
    }
  }

  val POST_USERS = new BusConfig1Field { def name = "post_users" }

  val GETALL_USERS = new BusConfig1Field { def name = "getall_users" }
}
